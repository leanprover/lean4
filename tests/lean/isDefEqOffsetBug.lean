class AddGroup (α : Type u) extends Add α, Zero α, Neg α where
  addAssoc : {a b c : α} → a + b + c = a + (b + c)
  zeroAdd  : {a : α} → 0 + a = a
  addZero  : {a : α} → a + 0 = a
  negAdd   : {a : α} → -a + a = 0

open AddGroup

theorem negZero [AddGroup α] : -(0 : α) = 0 := by
  rw [←addZero (a := -(0 : α)), negAdd]

theorem subZero [AddGroup α] {a : α} : a + -(0 : α) = a := by
  rw [← addZero (a := a)]
  rw [addAssoc]
  rw [negZero]
  rw [addZero]

theorem shouldFail [AddGroup α] : ((0 : α) + 0) = 0 :=
  (rfl) -- Error
