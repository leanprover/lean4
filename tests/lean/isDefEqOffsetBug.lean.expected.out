isDefEqOffsetBug.lean:19:2-19:5: error: type mismatch
  rfl
has type
  ?m = ?m : Prop
but is expected to have type
  0 + 0 = 0 : Prop
